File: backend/internal/model/coupon.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Coupon struct {
	ID            uuid.UUID `json:"id"`
	ProductID     uuid.UUID `json:"product_id"`
	Code          string    `json:"code"`
	DiscountType  string    `json:"discount_type"`  // "fixed" or "percentage"
	DiscountValue int64     `json:"discount_value"`
	MaxUses       int       `json:"max_uses"`
	UsedCount     int       `json:"used_count"`
	OncePerUser   bool      `json:"once_per_user"`
	ExpiresAt     time.Time `json:"expires_at"`
	CreatedAt     time.Time `json:"created_at"`

	// Joined fields
	ProductName string `json:"product_name,omitempty"`
}

type CreateCouponRequest struct {
	ProductID     uuid.UUID `json:"product_id"`
	Code          string    `json:"code"`
	DiscountType  string    `json:"discount_type" validate:"required,oneof=fixed percentage"`
	DiscountValue int64     `json:"discount_value" validate:"required,min=1"`
	MaxUses       int       `json:"max_uses" validate:"min=0"`
	OncePerUser   bool      `json:"once_per_user"`
	ExpiresAt     time.Time `json:"expires_at" validate:"required"`
}

type CouponUsage struct {
	ID             uuid.UUID `json:"id"`
	CouponID       uuid.UUID `json:"coupon_id"`
	UserID         uuid.UUID `json:"user_id"`
	SubscriptionID uuid.UUID `json:"subscription_id"`
	CreatedAt      time.Time `json:"created_at"`
}

type ValidateCouponRequest struct {
	Code      string    `json:"code" validate:"required"`
	ProductID uuid.UUID `json:"product_id" validate:"required"`
}

type ValidateCouponResponse struct {
	Valid          bool   `json:"valid"`
	DiscountAmount int64  `json:"discount_amount"`
	Message        string `json:"message"`
}

File: frontend/src/services/coupon.ts
import api from './api'

export interface Coupon {
  id: string
  product_id: string
  code: string
  discount_type: 'fixed' | 'percentage'
  discount_value: number
  max_uses: number
  used_count: number
  expires_at: string
  created_at: string
  product_name?: string
}

export interface CreateCouponRequest {
  code?: string
  discount_type: 'fixed' | 'percentage'
  discount_value: number
  max_uses: number
  expires_at: string
}

export interface CouponListResponse {
  coupons: Coupon[]
  total: number
  page: number
  limit: number
  total_pages: number
}

export interface CouponValidateResponse {
  valid: boolean
  discount_amount: number
  message: string
}

export const couponService = {
  async create(productId: string, data: CreateCouponRequest): Promise<Coupon> {
    const response = await api.post<Coupon>(`/products/${productId}/coupons`, data)
    return response.data
  },

  async listByProduct(productId: string, page = 1, limit = 20): Promise<CouponListResponse> {
    const response = await api.get<CouponListResponse>(`/products/${productId}/coupons`, { params: { page, limit } })
    return response.data
  },

  async listMy(page = 1, limit = 20): Promise<CouponListResponse> {
    const response = await api.get<CouponListResponse>('/me/coupons', { params: { page, limit } })
    return response.data
  },

  async delete(id: string): Promise<void> {
    await api.delete(`/coupons/${id}`)
  },

  async validate(code: string, productId: string): Promise<CouponValidateResponse> {
    const response = await api.post<CouponValidateResponse>('/coupons/validate', { code, product_id: productId })
    return response.data
  },
}

export default couponService
